### ISSUE TITLE
Lack of Input Validation in theano.gradient.verify_grad

### ISSUE THREAD

**Issue by rakeshbarman2790** [Theano/Theano#6817]
Issue with [theano.gradient.verify](http://theano.gradient.verify/)_grad Failing to Check Input Validity
The [theano.gradient.verify](http://theano.gradient.verify/)_grad function does not validate the compatibility of input shapes for the tested function. In the example below, input_1 and input_alpha are of mismatched shapes, but [theano.gradient.verify](http://theano.gradient.verify/)_grad fails to detect the issue and outputs None without any error. However, directly using these inputs with [theano.tensor.nnet.relu](http://theano.tensor.nnet.relu/) causes an error.

Problematic Code
import theano
import [theano.tensor](http://theano.tensor/) as T
import numpy as np

Custom activation function using Theano's ReLU
def custom_activation(x, alpha):
return [theano.tensor.nnet.relu](http://theano.tensor.nnet.relu/)(x, alpha)

Define symbolic tensors
x1 = [T.tensor](http://t.tensor/)3('x1')
alpha = [T.tensor](http://t.tensor/)4('alpha')

Compute output and gradients
output = custom_activation(x1, alpha)
loss = [T.sum](http://t.sum/)(output ** 2)
grad_x1, grad_alpha = [T.grad](http://t.grad/)(loss, [x1, alpha])

Define inputs with mismatched shapes
input_1 = [np.random.random](http://np.random.random/)((1, 2, 3)).astype('float32') # Shape: (1, 2, 3)
input_alpha = [np.random.random](http://np.random.random/)((1, 2, 1, 1)).astype('float32') # Shape: (1, 2, 1, 1)

Verify gradients
rng = [np.random.RandomState](http://np.random.randomstate/)(123)
print([theano.gradient.verify](http://theano.gradient.verify/)_grad(custom_activation, pt=[input_1, input_alpha], rng=rng))

Explanation of the Issue
Shape Mismatch: The inputs input_1 and input_alpha must have compatible shapes for broadcasting when passed to [theano.tensor.nnet.relu](http://theano.tensor.nnet.relu/). In this case, they are not compatible.
No Error from verify_grad: Despite the incompatibility, [theano.gradient.verify](http://theano.gradient.verify/)_grad does not validate the input shapes before proceeding, leading to an incorrect None output instead of raising an error.
Error in [theano.tensor.nnet.relu](http://theano.tensor.nnet.relu/): If custom_activation is used directly, the underlying [theano.tensor.nnet.relu](http://theano.tensor.nnet.relu/) function correctly raises an error due to the shape [mismatch.rewrite](http://mismatch.rewrite/) this
ChatGPT said:
ChatGPT
Lack of Input Validation in [theano.gradient.verify](http://theano.gradient.verify/)_grad

Issue Description
The [theano.gradient.verify](http://theano.gradient.verify/)_grad function does not validate the compatibility of input shapes for the tested function. This can result in misleading behavior where the function does not raise an error even when inputs have incompatible shapes. However, directly using the same inputs with functions like [theano.tensor.nnet.relu](http://theano.tensor.nnet.relu/) causes an appropriate shape mismatch error.

Problematic Code Example
import theano
import [theano.tensor](http://theano.tensor/) as T
import numpy as np

# Custom activation function using Theano's ReLU
def custom_activation(x, alpha):
    return [theano.tensor.nnet.relu](http://theano.tensor.nnet.relu/)(x, alpha)

# Define symbolic tensors
x1 = [T.tensor](http://t.tensor/)3('x1')
alpha = [T.tensor](http://t.tensor/)4('alpha')

# Compute output and gradients
output = custom_activation(x1, alpha)
loss = [T.sum](http://t.sum/)(output ** 2)
grad_x1, grad_alpha = [T.grad](http://t.grad/)(loss, [x1, alpha])

# Define inputs with mismatched shapes
input_1 = [np.random.random](http://np.random.random/)((1, 2, 3)).astype('float32')  # Shape: (1, 2, 3)
input_alpha = [np.random.random](http://np.random.random/)((1, 2, 1, 1)).astype('float32')  # Shape: (1, 2, 1, 1)

# Verify gradients
rng = [np.random.RandomState](http://np.random.randomstate/)(123)
print([theano.gradient.verify](http://theano.gradient.verify/)_grad(custom_activation, pt=[input_1, input_alpha], rng=rng))
Explanation of the Issue
Shape Mismatch

The inputs input_1 and input_alpha must have compatible shapes for broadcasting in the function [theano.tensor.nnet.relu](http://theano.tensor.nnet.relu/). In this example, their shapes (1, 2, 3) and (1, 2, 1, 1) are incompatible.
Behavior of verify_grad

The [theano.gradient.verify](http://theano.gradient.verify/)_grad function does not validate input shape compatibility. Instead of detecting the issue and raising an error, it outputs None.
Error Handling in [theano.tensor.nnet.relu](http://theano.tensor.nnet.relu/)

When custom_activation is used directly (bypassing verify_grad), the underlying [theano.tensor.nnet.relu](http://theano.tensor.nnet.relu/) function correctly identifies the shape mismatch and raises an error.
Root Cause
The issue arises because [theano.gradient.verify](http://theano.gradient.verify/)_grad does not check whether the provided inputs are compatible with the function under test. It relies on the tested function to handle any such validation, which might not always occur during gradient verification.

Suggested Fix or Workaround
To avoid such issues, explicitly validate input shapes before passing them to verify_grad. For example:

# Ensure inputs have compatible shapes for broadcasting
if not [np.array](http://np.array/)_equal([np.broadcast](http://np.broadcast/)_shapes([input_1.shape](http://input_1.shape/), [input_alpha.shape](http://input_alpha.shape/)), [input_1.shape](http://input_1.shape/)):
    raise ValueError("Input shapes are not compatible for broadcasting")
Alternatively, modify custom_activation to validate inputs:

def custom_activation(x, alpha):
    if not [np.array](http://np.array/)_equal([np.broadcast](http://np.broadcast/)_shapes([x.shape](http://x.shape/), [alpha.shape](http://alpha.shape/)), [x.shape](http://x.shape/)):
        raise ValueError("Input shapes are not compatible for broadcasting")
    return [theano.tensor.nnet.relu](http://theano.tensor.nnet.relu/)(x, alpha)
Expected Behavior
Without Input Validation: verify_grad fails silently and outputs None.
With Input Validation: An informative error is raised when the inputs are incompatible, both during direct usage and gradient verification.
By incorporating input validation, such issues can be mitigated effectively.
